handlers/user.go

```diff
@@ -2078,7 +2079,43 @@ func isValidSubscriptionFilter(eventName types.EventName, filter string) bool {
 		isClient = true
 	}
 
-	return len(filter) == 0 || isPkey || isClient
+	isValidMachine := false
+	if types.IsMachineNotification(eventName) {
+		machines, err := db.BigtableClient.GetMachineMetricsMachineNames(userID)
+		if err != nil {
+			return false, errors.Wrap(err, "can not get users machines from bigtable for validation")
+		}
+		for _, userMachineName := range machines {
+			if userMachineName == filter {
+				isValidMachine = true
+				break
+			}
+		}
+
+		// While the above works fine for active machines (adding a new notification to an active machine)
+		// It does not work for a machine that is offline and where the user wants to unsubscribe from this machine.
+		// So check the db for any machine names as well
+		if !isValidMachine {
+			machines := make([]string, 0)
+			err = db.FrontendWriterDB.Select(&machines, `
+				select event_filter
+				from users_subscriptions 
+				where user_id = $1 AND event_name = ANY($2)
+			`, userID, pq.Array(types.MachineEvents))
+			if err != nil {
+				return false, errors.Wrap(err, "can not get event_filters from db for validation")
+			}
+
+			for _, machineName := range machines {
+				if machineName == filter {
+					isValidMachine = true
+					break
+				}
+			}
+		}
+	}
```

Comment:
**Issue**: IMO this doesn't fit here, as this function is supposed to be a simple input validation and not a check against the DB state. You should rather extract all logic regarding machine metrics into it's own function. This would also allow for better API error responses, as error code `404` would be more appropriate in case of not finding the machine.

Reply:
Improved api error